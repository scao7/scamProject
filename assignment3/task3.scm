(define (node v pa ch)
    (define val v)
    (define parent pa)
    (define child ch)

    (define (set-parent p)
        (set 'parent p)
        )


    (define (set-child c)
        (set 'child c)
        )


    (define (set-val va)
        (set 'val va)
        )

    
    (define (get-parent)
        parent
        )

    
    (define (get-child)
        child
        )


    (define (get-val)
        val
        )


    this
    )


(define (deque)
    (define stuff nil)
    (define size 0)
    (define tail nil)


    (define (enqueueFront val)
        (define n (node val nil nil))
        (if (null? stuff)
            (begin
                (set 'size 1)
                (set 'stuff n)
                (set 'tail n)
                )
            (begin
                ((n'set-child) stuff)
                ((stuff'set-parent) n)
                (set 'stuff n)
                (++ size)
                )
            )
        val
        )


    (define (enqueueBack val)
        (define n (node val nil nil))
        (if (null? stuff)
            (begin
                (set 'size 1)
                (set 'stuff n)
                (set 'tail n)
                )
            (begin
                ((n'set-parent) tail)
                ((tail'set-child) n)
                (set 'tail n)
                (++ size)
                )
            )
        val
        )


    (define (enqueueIndex ind val)
        (define (e-iter src count in va)
            (if (= count in)
                (begin
                    (define c ((src'get-child)))
                    (define n (node va src c))
                    ((src'set-child) n)
                    (if (= in size)
                        (set 'tail n)
                        ((c'set-parent) n)
                        )
                    )
                (e-iter ((src'get-child)) (+ count 1) in va)
                )
            )

        (if (= ind 0)
            (enqueueFront val)
            (begin
                (e-iter stuff 1 ind val)
                (++ size)
                )
            )
        val
        )


    (define (dequeueFront)
        (define v ((stuff'get-val)))
        (set 'stuff ((stuff'get-child)))
        (-- size)
        (if (= size 0)
            (set 'tail nil)
            ((stuff'set-parent) nil)
            )
        v
        )


    (define (dequeueBack)
        (define v ((tail'get-val)))
        (set 'tail ((tail'get-parent)))
        (-- size)
        (if (= size 0)
            (set 'stuff nil)
            ((tail'set-child) nil)
            )
        v
        )


    (define (dequeueIndex ind)
        (define (d-iter src ind count)
            (if (= ind count)
                (begin
                    (define c ((src'get-child)))
                    (define p ((src'get-parent)))
                    (define v ((src'get-val)))
                    ((c'set-parent) p)
                    (if (!= count 0)
                        ((p'set-child) c)
                        (set 'stuff c)
                        )
                    v
                    )
                (d-iter ((src'get-parent)) ind (- count 1))
                )
            )
        

        (if (= ind (- size 1))
            (dequeueBack)
            (begin
                (define v (d-iter ((tail'get-parent)) ind (- size 2)))
                (-- size)
                v
                )
            )
        )

    
    (define (display)
        (define (d-iter src)
            (if (not (null? src))
                (begin 
                    (print ((src'get-val)))
                    (if (not (null? ((src'get-child))))
                        (begin 
                            (print ",")
                            (d-iter ((src'get-child)))
                            )
                        )
                    )
                )
            )

        (print "[")
        (d-iter stuff)
        (print "]")
        )


    (define (peekFront)
        ((stuff'get-val))
        )

    
    (define (peekBack)
        ((tail'get-val))
        )
        

    this
    )


(define (main)
    (define env this)
    (define (iter expr)
         (if (not (eof?)) (begin (eval expr env) (iter (readExpr))))
         )
    (setPort (open (getElement ScamArgs 1) 'read))
    (iter (readExpr))
    )