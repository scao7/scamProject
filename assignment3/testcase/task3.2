(define dq (deque))
((dq'enqueueIndex) 0 60)
((dq'enqueueIndex) 1 38)
((dq'enqueueIndex) 2 96)
((dq'enqueueIndex) 2 61)
((dq'enqueueIndex) 2 14)
((dq'enqueueIndex) 5 39)
((dq'enqueueIndex) 5 22)
((dq'enqueueIndex) 1 78)
((dq'enqueueIndex) 8 47)
((dq'enqueueIndex) 1 80)
((dq'display))
(println)
((dq'dequeueIndex) 6)
((dq'display))
(println)
((dq'dequeueIndex) 7)
((dq'display))
(println)
((dq'dequeueIndex) 5)
((dq'display))
(println)
((dq'dequeueIndex) 4)
((dq'display))
(println)
((dq'dequeueIndex) 4)
((dq'display))
(println)
((dq'dequeueIndex) 4)
((dq'display))
(println)
((dq'dequeueIndex) 3)
((dq'display))
(println)
((dq'dequeueIndex) 1)
((dq'display))
(println)
((dq'dequeueIndex) 1)
((dq'display))
(println)
((dq'dequeueIndex) 0)
((dq'display))
(println)
