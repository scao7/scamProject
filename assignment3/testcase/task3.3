(define dq (deque))
((dq'enqueueIndex) 0 26)
((dq'enqueueIndex) 1 10)
((dq'enqueueIndex) 0 24)
((dq'enqueueIndex) 2 11)
((dq'enqueueIndex) 2 44)
((dq'enqueueIndex) 5 48)
((dq'enqueueIndex) 1 67)
((dq'enqueueIndex) 0 40)
((dq'enqueueIndex) 6 12)
((dq'enqueueIndex) 0 44)
((dq'display))
(println)
((dq'dequeueIndex) 3)
((dq'display))
(println)
((dq'dequeueIndex) 7)
((dq'display))
(println)
((dq'dequeueIndex) 1)
((dq'display))
(println)
((dq'dequeueIndex) 6)
((dq'display))
(println)
((dq'dequeueIndex) 3)
((dq'display))
(println)
((dq'dequeueIndex) 3)
((dq'display))
(println)
((dq'dequeueIndex) 1)
((dq'display))
(println)
((dq'dequeueIndex) 1)
((dq'display))
(println)
((dq'dequeueIndex) 1)
((dq'display))
(println)
((dq'dequeueIndex) 0)
((dq'display))
(println)
((dq'display))
(println)
((dq'enqueueIndex) 0 69)
((dq'enqueueIndex) 1 27)
((dq'enqueueIndex) 1 85)
((dq'enqueueIndex) 1 43)
((dq'enqueueIndex) 4 93)
((dq'enqueueIndex) 4 83)
((dq'enqueueIndex) 2 10)
((dq'enqueueIndex) 3 51)
((dq'enqueueIndex) 3 27)
((dq'enqueueIndex) 4 45)
((dq'display))
(println)
((dq'dequeueIndex) 5)
((dq'display))
(println)
((dq'dequeueIndex) 7)
((dq'display))
(println)
((dq'dequeueIndex) 1)
((dq'display))
(println)
((dq'dequeueIndex) 5)
((dq'display))
(println)
((dq'dequeueIndex) 4)
((dq'display))
(println)
((dq'dequeueIndex) 3)
((dq'display))
(println)
((dq'dequeueIndex) 1)
((dq'display))
(println)
((dq'dequeueIndex) 0)
((dq'display))
(println)
((dq'dequeueIndex) 0)
((dq'display))
(println)
((dq'dequeueIndex) 0)
((dq'display))
(println)
