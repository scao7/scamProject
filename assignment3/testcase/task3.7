(include "random.lib")
(randomSeed 1096)
(define dq (deque))
(define (upfront dq n)
    (define value (randomRange 10 100))
    (cond
        ((= n 0)
            ((dq'enqueueFront) value)
            )
        ((< (dq'size) 10)
            ((dq'enqueueFront) value)
            (upfront dq (- n 1))
            )
        (else
            ((dq'enqueueIndex) (randomRange 2 10) value)
            (upfront dq (- n 1))
            )
        )
    )
(define (upback dq n)
    (define value (randomRange 10 100))
    (cond
        ((= n 0)
            ((dq'enqueueBack) value)
            )
        (else
            ((dq'enqueueBack) value)
            (upback dq (- n 1))
            )
        )
    )
(define (downfront dq n)
    (cond
        ((= n 0)
            ((dq'dequeueFront))
            )
        (else
            ((dq'dequeueFront))
            (downfront dq (- n 1))
            )
        )
    )
(upfront dq 2531)
(upback dq 2567)
(downfront dq 4950)
((dq'display)) (println)
