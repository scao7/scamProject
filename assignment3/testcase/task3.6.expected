[71,17,68,20,40,81,13,43,76,93,77,81,81,27,42,67,28,97,41,32,...]
[27,64,79,40,94,78,37,18,39,93,49,51,26,37,63,26,16,17,28,63,...]
